File: frontend/src/services/languageService.ts
import api from './api'
import type { Language } from '../types'

export const languageService = {
  getAll: async (edition?: string): Promise<Language[]> => {
    const { data } = await api.get('/languages', {
      params: edition ? { edition } : {},
    })
    return data
  },

  getByCharacter: async (characterId: number): Promise<Language[]> => {
    const { data } = await api.get(`/characters/${characterId}/languages`)
    return data
  },

  add: async (characterId: number, languageId: number): Promise<void> => {
    await api.post(`/characters/${characterId}/languages/${languageId}`)
  },

  remove: async (characterId: number, languageId: number): Promise<void> => {
    await api.delete(`/characters/${characterId}/languages/${languageId}`)
  },
}

File: backend/internal/service/language_service.go
package service

import "rpg-manager/internal/domain"

type languageRepo interface {
	GetAll(edition string) ([]domain.Language, error)
	GetByCharacter(characterID uint) ([]domain.Language, error)
	Add(characterID, languageID uint) error
	Remove(characterID, languageID uint) error
}

type LanguageService struct{ repo languageRepo }

func NewLanguageService(repo languageRepo) *LanguageService { return &LanguageService{repo: repo} }

func (s *LanguageService) GetAll(edition string) ([]domain.Language, error) {
	return s.repo.GetAll(edition)
}
func (s *LanguageService) GetByCharacter(id uint) ([]domain.Language, error) {
	return s.repo.GetByCharacter(id)
}
func (s *LanguageService) Add(charID, languageID uint) error {
	return s.repo.Add(charID, languageID)
}
func (s *LanguageService) Remove(charID, languageID uint) error {
	return s.repo.Remove(charID, languageID)
}
